modules/frontend/queue/queue.go

```diff
@@ -159,16 +161,30 @@ FindQueue:
 	queue, userID, idx := q.queues.getNextQueueForQuerier(last.last)
 	last.last = idx
 	if queue != nil {
+		guaranteedInQueue := requestedCount
 		// this is all threadsafe b/c all users queues are blocked by q.mtx
 		if len(queue) < requestedCount {
-			requestedCount = len(queue)
+			guaranteedInQueue = len(queue)
 		}
 
-		// Pick next requests from the queue.
-		batchBuffer = batchBuffer[:requestedCount]
-		for i := 0; i < requestedCount; i++ {
+		totalWeight := 0
+		actuallyInBatch := 0
+		for i := 0; i < guaranteedInQueue; i++ {
 			batchBuffer[i] = <-queue
+			actuallyInBatch++
+
+			// PRTODO: add tests that take weight into account
+			weight := batchBuffer[i].Weight()
+			if weight <= 0 {
+				weight = 1
+			}
+			totalWeight += weight
+
+			if totalWeight >= requestedCount {
```

Comment:
I think this makes sense.  I suppose what we're saying here is that we request of this batch a certain high water mark of work that we're willing to take, and the weight increases the notion of complexity for a single item above this threshold.  Implicitly here I suppose is that weight and requestedCount are of the same unit of measure.

Reply:
> Implicitly here I suppose is that weight and requestedCount are of the same unit of measure.

yes! currently all jobs fill a single "slot" in the batch. the "weight" is basically just making it fill more slots.